packages/arb-avm-cpp/data_storage/src/datastorage.cpp

```diff
@@ -274,31 +277,22 @@ ConcurrentCounter::ConcurrentCounter(const DataStorage* _storage)
     }
 }
 
-ConcurrentCounter::~ConcurrentCounter() {
+TryDbLockShared::~TryDbLockShared() {
     if (storage == nullptr) {
+        // Don't do anything if counter was moved
         return;
     }
-    auto new_counter = storage->concurrent_database_access_counter.fetch_sub(1);
-    assert(new_counter >= 0);
-    (void)new_counter;  // silence unused variable warning on release builds
-                        // where assert is disabled
-}
-
-ConcurrentCounter ConcurrentCounter::Get(const DataStorage* storage) {
-    if (storage->shutting_down) {
-        throw DataStorage::shutting_down_exception();
-    }
-    return ConcurrentCounter(storage);
+    storage->concurrent_database_access_counter--;
+    assert(storage->concurrent_database_access_counter.load() >= 0);
```

Comment:
No need for two reads, we can just get the return value from the `--` line

Reply:
:+1: